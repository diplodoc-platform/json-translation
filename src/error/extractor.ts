class ExtractorError extends Error {
    constructor(message: string) {
        super(message);
        this.name = 'ExtractorError';

        Error.captureStackTrace(this, this.constructor);
    }
}

export {ExtractorError};
export default {ExtractorError};
