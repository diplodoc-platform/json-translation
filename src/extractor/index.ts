export * from './extractor';

export type * from './extractor';
