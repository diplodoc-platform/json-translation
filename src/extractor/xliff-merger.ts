import {Builder, Parser} from 'xml2js';

import {ExtractParameters} from '@diplodoc/markdown-translation';

export type XLIFFMergerParameters = {
    sourceLanguageLocale?: ExtractParameters['source'];
    targetLanguageLocale?: ExtractParameters['target'];
    dataType?: string;
    contentName?: string;
    skeletonName?: string;
    xmlNamespace?: string;
    xmlVersion?: string;
};

class XLIFFMerger {
    private xmlNamespace_: string;
    private xmlVersion_: string;

    private sourceLanguageLocale_: ExtractParameters['source'];
    private targetLanguageLocale_: ExtractParameters['target'];

    private dataType_: string;
    private contentName_: string;
    private skeletonName_: string;

    private parser: Parser;
    private builder: Builder;

    constructor(parameters: XLIFFMergerParameters) {
        const {
            sourceLanguageLocale,
            targetLanguageLocale,
            dataType,
            contentName,
            skeletonName,
            xmlNamespace,
            xmlVersion,
        } = parameters;

        this.sourceLanguageLocale_ = sourceLanguageLocale ?? {language: 'ru', locale: 'RU'};
        this.targetLanguageLocale_ = targetLanguageLocale ?? {language: 'en', locale: 'US'};

        this.dataType_ = dataType ?? 'markdown';
        this.contentName_ = contentName ?? 'json-translation-content';
        this.skeletonName_ = skeletonName ?? 'json-translation-skeleton';
        this.xmlNamespace_ = xmlNamespace ?? 'urn:oasis:names:tc:xliff:document:1.2';
        this.xmlVersion_ = xmlVersion ?? '1.2';

        this.parser = new Parser();
        this.builder = new Builder();
    }

    async merge(xliffs: Record<string, string>): Promise<string> {
        const merged = [];

        for (const [i, xliff] of Object.entries(xliffs)) {
            if (!xliff.length) {
                continue;
            }

            const parsed = await this.parser.parseStringPromise(xliff);
            const transUnits = parsed?.xliff?.file[0]?.body[0]['trans-unit'];
            if (!transUnits) {
                continue;
            }

            for (let j = 0; j < transUnits.length; j++) {
                transUnits[j].$.id = `${i}-${transUnits[j].$.id}`;
            }

            merged.push(...transUnits);
        }

        const document = this.xliffTemplate(merged);
        const rendered = this.builder.buildObject(document);

        return rendered;
    }

    private xliffTemplate(transUnits: unknown[]) {
        return {
            xliff: {
                $: {
                    xmlns: this.xmlNamespace_,
                    version: this.xmlVersion_,
                },
                file: [
                    {
                        $: {
                            original: this.contentName_,
                            'source-language': `${this.sourceLanguageLocale_.language}-${this.sourceLanguageLocale_.locale}`,
                            'target-language': `${this.targetLanguageLocale_.language}-${this.targetLanguageLocale_.locale}`,
                            datatype: this.dataType_,
                        },
                        header: [
                            {
                                skeleton: [
                                    {
                                        'external-file': [
                                            {
                                                $: {
                                                    href: this.skeletonName_,
                                                },
                                            },
                                        ],
                                    },
                                ],
                            },
                        ],
                        body: [{'trans-unit': transUnits}],
                    },
                ],
            },
        };
    }
}

export {XLIFFMerger};
export default {XLIFFMerger};
