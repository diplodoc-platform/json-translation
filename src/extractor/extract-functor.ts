import {ExtractParameters, extract} from '@diplodoc/markdown-translation';

import {IFunctor} from 'src/functor';

export type ExtractFunctorParameters = {
    initialCounter: number;
};

class ExtractFunctor implements IFunctor<string> {
    private id_: number;
    private skeletons_: Record<string, string>;
    private xliffs_: Record<string, string>;
    private markdownTranslationArguments: ExtractParameters;

    get xliffs() {
        return this.xliffs_;
    }

    get skeletons() {
        return this.skeletons_;
    }

    constructor(parameters: ExtractFunctorParameters) {
        const {initialCounter} = parameters;

        this.id_ = initialCounter;
        this.skeletons_ = {};
        this.xliffs_ = {};

        this.markdownTranslationArguments = {
            source: {
                language: 'ru',
                locale: 'RU' as const,
            },
            target: {
                language: 'en',
                locale: 'US' as const,
            },
        };
    }

    call(str: string) {
        const key = `${this.id_}`;

        const extractArguments = {
            ...this.markdownTranslationArguments,
            markdown: str,
            markdownPath: `${this.id_}.md`,
            skeletonPath: `${this.id_}.skl.md`,
        };

        const {skeleton, xlf} = extract(extractArguments);

        this.skeletons_[key] = skeleton;
        this.xliffs_[key] = xlf;

        return this.skeleton();
    }

    private skeleton() {
        return `%%%${this.id_++}%%%`;
    }
}

export {ExtractFunctor};
export default {ExtractFunctor};
