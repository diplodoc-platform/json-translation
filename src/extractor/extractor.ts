import {JSONSchema7} from 'json-schema';

import {JSONTranslationExtractParameters, JSONTranslationExtractResult} from 'src/api';
import {JSONRefResolver, JSONTraversal} from 'src/json';
import {ExtractorError} from 'src/error/extractor';
import {ExtractFunctor} from './extract-functor';
import {XLIFFMerger} from './xliff-merger';

const SCHEMA_KEYWORD = 'translatable';

class Extractor {
    schema: JSONSchema7;
    data: object;
    schemaKeyword: string;

    extractFunctor: ExtractFunctor;
    jsonRefResolver: JSONRefResolver;
    jsonTraversal: JSONTraversal;
    xliffMerger: XLIFFMerger;

    constructor(parameters: JSONTranslationExtractParameters) {
        this.schema = parameters.schema;
        this.data = parameters.data;
        this.schemaKeyword = parameters.schemaKeyword ?? SCHEMA_KEYWORD;

        // dependencies
        this.extractFunctor = new ExtractFunctor({initialCounter: 1});
        this.jsonRefResolver = new JSONRefResolver({resolveCircular: true});
        const jsonTraversalArguments = {
            schema: this.schema,
            schemaKeyword: this.schemaKeyword,
        };
        this.jsonTraversal = new JSONTraversal(jsonTraversalArguments);
        this.xliffMerger = new XLIFFMerger({});
    }

    async extract(): Promise<JSONTranslationExtractResult> {
        try {
            this.schema = await this.jsonRefResolver.resolve(this.schema);
        } catch (err: unknown) {
            throw new ExtractorError('failed to resolve schema refs');
        }

        const extractHandler = this.extractFunctor.call.bind(this.extractFunctor);
        this.jsonTraversal.addNodeHandler('string', this.schemaKeyword, extractHandler);

        try {
            this.jsonTraversal.traverse(this.data);
        } catch (err: unknown) {
            throw new ExtractorError('failed to traverse data using schema');
        }

        let skeleton;
        let xliff;

        try {
            skeleton = this.extractFunctor.skeletons;
            xliff = await this.xliffMerger.merge(this.extractFunctor.xliffs);
        } catch (err: unknown) {
            throw new ExtractorError('failed to prepare xliff');
        }

        return {
            skeleton,
            xliff,
        };
    }
}

export {Extractor};
export default {Extractor};
