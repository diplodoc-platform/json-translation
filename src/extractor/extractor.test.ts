import {Extractor} from './extractor';

const schema = require('src/mvp/__fixtures__/schema.json');
const data = require('src/mvp/__fixtures__/data.json');

describe('extractor', () => {
    it('generates skeleton and xliff', async () => {
        const extractorArguments = {
            schema,
            data,
        };

        const extractor = new Extractor(extractorArguments);
        const {skeleton, xliff} = await extractor.extract();

        expect(skeleton).toMatchSnapshot();
        expect(xliff).toMatchSnapshot();
    });
});
