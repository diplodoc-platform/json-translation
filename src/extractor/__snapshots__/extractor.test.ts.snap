// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractor generates skeleton and xliff 1`] = `
Object {
  "1": "%%%1%%%",
  "10": "%%%1%%%",
  "11": "%%%1%%% %%%2%%%",
  "12": "%%%1%%% %%%2%%% %%%3%%%",
  "13": "%%%1%%%",
  "14": "%%%1%%%",
  "15": "%%%1%%% %%%2%%%",
  "16": "%%%1%%% %%%2%%%",
  "17": "%%%1%%%",
  "18": "%%%1%%% %%%2%%%",
  "19": "%%%1%%%",
  "2": "%%%1%%%",
  "20": "%%%1%%%",
  "21": "%%%1%%%",
  "3": "%%%1%%%",
  "4": "%%%1%%%",
  "5": "%%%1%%%",
  "6": "%%%1%%%",
  "7": "%%%1%%%",
  "8": "%%%1%%%",
  "9": "%%%1%%%",
}
`;

exports[`extractor generates skeleton and xliff 2`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" standalone=\\"yes\\"?>
<xliff xmlns=\\"urn:oasis:names:tc:xliff:document:1.2\\" version=\\"1.2\\">
  <file original=\\"json-translation-content\\" source-language=\\"ru-RU\\" target-language=\\"en-US\\" datatype=\\"markdown\\">
    <header>
      <skeleton>
        <external-file href=\\"json-translation-skeleton\\"/>
      </skeleton>
    </header>
    <body>
      <trans-unit id=\\"1-1\\">
        <source>DataLens Festival</source>
      </trans-unit>
      <trans-unit id=\\"2-1\\">
        <source>Yandex DataLens Festival</source>
      </trans-unit>
      <trans-unit id=\\"3-1\\">
        <source>Для кого этот фестиваль</source>
      </trans-unit>
      <trans-unit id=\\"4-1\\">
        <source>Вас ждут</source>
      </trans-unit>
      <trans-unit id=\\"5-1\\">
        <source>Зачем участвовать</source>
      </trans-unit>
      <trans-unit id=\\"6-1\\">
        <source>Программа</source>
      </trans-unit>
      <trans-unit id=\\"7-1\\">
        <source>hello</source>
      </trans-unit>
      <trans-unit id=\\"8-1\\">
        <source>Неделя 2: корпоративная аналитика в Yandex Cloud</source>
      </trans-unit>
      <trans-unit id=\\"9-1\\">
        <source>hello</source>
      </trans-unit>
      <trans-unit id=\\"10-1\\">
        <source>Частые вопросы</source>
      </trans-unit>
      <trans-unit id=\\"11-1\\">
        <source>Кто может участвовать в фестивале?</source>
      </trans-unit>
      <trans-unit id=\\"11-2\\">
        <source>Что для этого нужно?</source>
      </trans-unit>
      <trans-unit id=\\"12-1\\">
        <source>В фестивале могут участвовать все желающие аналитики, руководители и тимлиды из любой точки мира.</source>
      </trans-unit>
      <trans-unit id=\\"12-2\\">
        <source>Мероприятие проходит онлайн.</source>
      </trans-unit>
      <trans-unit id=\\"12-3\\">
        <source>Участие бесплатное.</source>
      </trans-unit>
      <trans-unit id=\\"13-1\\">
        <source>Нужна ли регистрация, чтобы слушать онлайн-уроки?</source>
      </trans-unit>
      <trans-unit id=\\"14-1\\">
        <source>Мы рекомендуем зарегистрироваться, чтобы получить доступ ко всем материалам онлайн-фестиваля, а также не упустить возможность принять участие в конкурсе.</source>
      </trans-unit>
      <trans-unit id=\\"15-1\\">
        <source>Я не аналитик и не руководитель.</source>
      </trans-unit>
      <trans-unit id=\\"15-2\\">
        <source>Можно мне на фестиваль?</source>
      </trans-unit>
      <trans-unit id=\\"16-1\\">
        <source>Программа ориентирована на аудиторию, которая знакома с принципами работы BI или имела опыт в построении системы бизнес-аналитики.</source>
      </trans-unit>
      <trans-unit id=\\"16-2\\">
        <source>
          В рамках этого мероприятия мы не будем рассматривать базовые понятия, но вы можете присоединиться к другим .
          <g ctype=\\"x-link_text_part\\" equiv-text=\\"[]\\">мероприятиям Yandex Cloud</g>
          <g ctype=\\"x-link_attributes_part\\" equiv-text=\\"()\\">
            <x ctype=\\"x-link_attributes_href\\" equiv-text=\\"https://cloud.yandex.ru/events\\"/>
          </g>
        </source>
      </trans-unit>
      <trans-unit id=\\"17-1\\">
        <source>Что я получу после фестиваля?</source>
      </trans-unit>
      <trans-unit id=\\"18-1\\">
        <source>Вы сможете  принять участие в конкурсе на лучшее выполнение задания.</source>
      </trans-unit>
      <trans-unit id=\\"18-2\\">
        <source>Сертификатов и дипломов за посещение мероприятий фестиваля не предусмотрено.</source>
      </trans-unit>
      <trans-unit id=\\"19-1\\">
        <source>Будут ли записи онлайн‑уроков?</source>
      </trans-unit>
      <trans-unit id=\\"20-1\\">
        <source>
          Да, вы сможете посмотреть записи в личном кабинете и после окончания фестиваля — на .
          <g ctype=\\"x-link_text_part\\" equiv-text=\\"[]\\">Youtube-канале Yandex Cloud</g>
          <g ctype=\\"x-link_attributes_part\\" equiv-text=\\"()\\">
            <x ctype=\\"x-link_attributes_href\\" equiv-text=\\"https://www.youtube.com/c/YandexCloudPlatform/\\"/>
          </g>
        </source>
      </trans-unit>
      <trans-unit id=\\"21-1\\">
        <source>
          Регистрируйтесь на фестиваль,  участие бесплатно
          <x ctype=\\"x-html_inline\\" equiv-text=\\"&lt;br>\\"/>
        </source>
      </trans-unit>
    </body>
  </file>
</xliff>"
`;
